libsystemd: remove optional OpenSSL linking, use khash unconditionally

bluca opened the pull request:
The shared libraries are licensed under LGPL-2.1-or-later. But linking them
against OpenSSL 3+, which is Apache2 license, forces them to be distributed
under LGPL-3.0-or-later, as (L)GPL-2 is not compatible with Apache2. This
could be quite disruptive for users linking against the shared library,
as a simple distribution upgrade moving OpenSSL to 3.0 will cause the
licensing situation of their programs to suddenly and unexpectedly change.

Whether right or wrong, GPL-3 is problematic in many settings, and many
projects go to extraordinary lengths to avoid it (eg: the Yocto project has
a flag to remove any and all GPL-3 sources from an image).

To avoid any ambiguity, always use khash in the sd-id128 APIs implementation.

I've left out considerations about the effect of the licenses from the README, and simply stated the facts about linkage, following reviews on similar changes.

The situation as far as I can see is that Debian and Ubuntu never enabled the openssl build flag, but Fedora does. So this is a small material change for Fedora, and a no-op for Debian/Ubuntu, so it should be sufficiently battle-tested and safe to do.

Yes using khash is not nice, but dealing with angry lawyers is worse.

poettering commented:
Uh, what? please, let's not slow down Fedora unnecessarily. If some distros don't want to go for slow khash instead of openssl for this, great, but let's not make that mandatory.

I am not convinced at all this is desirable. This cements one reading of the legal situation, and one I think is not universally accepted.

bluca commented:
> So you are basically saying that switching Fedora OpenSSL 3 actually _creates_ new license issues instead of removing them? Because before Fedora considered OpenSSL to be part of the core OS and thus circumvented the system lib wording of LGPL. But with the new license that doesn#t work anymore and OpenSSL 3 suddenly will create problems for software relying on this?

That is my personal interpretation, yes. The fact that they went through all the trouble of relicensing to solve the incompatibility with the GPL, just to pick the *ONE* permissive license that is still clearly incompatible with a version of the GPL is simply mind-boggling to me, but alas that's where we are. It is a step forward of course: GPL3+ and Apache2 are clearly and unambiguously compatible, so that's good at least - when that's possible, there's a clear way out.

Anyway, the issue is always uncertainty - where there's license doubts there's trouble. There's enormous value in having a situation that is crystal clear and not subject to any possible alternative interpretation. Which is where I'm trying to get us toward.

Here's a fun one, that explains why I'm so happy that I'm not a lawyer: exceptions are not transitive. Even if we had one of the usual "OpenSSL + GPL clauses" acked by all contributors applied to libsystemd.so, it would _only_ apply to the library itself. It would *NOT* apply to any external program linking to libsystemd.so. I recently learned this. Fun eh? Now if this is true, what does this mean for a vendor that claims "system lib clause" exception? Assuming it's valid, it's clear it applies to said vendor's first-party programs that link to libssl. Does it apply to third party programs that link to a vendor library that links to libssl? Does the third-party license situation change depending on which distro the program is deployed on, even if it's the same identical binary a libsystemd.so builds are ABI-compatible, depending on whether the vendor claims the system exception or not? I haven't the faintest clue. Actually I don't want to know, and I'd really like our users not to have to ponder about that too.

poettering commented:
So, we had a brief discussion elsewhere and the thing is that the system library exception isn't affected by the openssl license change: if you your distro was OK with assuming that openssl was a system lib before then it can assume that it is after too, the license change doesn't affect the assumption. Fedora/RHEL famously assume openssl is a system lib, hence it's find there to continue to link against openssl from any gpl code.

Hence, I am pretty sure we should not make this change: what was OK by fedora/rhel rules before should remain OK for them. Hence, the option to link directly to openssl should be kept. If debian/others disagree that is fine, and they can use khash, but I am not convinced we should from upstream mandate these changes for rhel/fedora given they never had a problem with assuming the system extension rule applying here.

Here's another idea we had: we currently embed a copy of SHA into sd-boot, because we cannot really link to anything there. It's copied out of glibc, should have OK licensing for all out users hence. Maybe we should move that from sd-boot into src/fundamental/ so that we can compile it for the userspace version too. If we embed a version of that anyway, we might as well use it at multiple places. If we did this we can drop the khash stuff, which Id like very much.

bluca commented:
> So, we had a brief discussion elsewhere and the thing is that the system library exception isn't affected by the openssl license change: if you your distro was OK with assuming that openssl was a system lib before then it can assume that it is after too, the license change doesn't affect the assumption. Fedora/RHEL famously assume openssl is a system lib, hence it's find there to continue to link against openssl from any gpl code.
> 
> Hence, I am pretty sure we should not make this change: what was OK by fedora/rhel rules before should remain OK for them. Hence, the option to link directly to openssl should be kept. If debian/others disagree that is fine, and they can use khash, but I am not convinced we should from upstream mandate these changes for rhel/fedora given they never had a problem with assuming the system extension rule applying here.

Thing is, as mentioned these exceptions are likely not transitive - while it works for executables shipped by rhel/fedora, it doesn't necessarily apply to third parties shipping their own executables. That's why the various tools and daemons in our repo are fine, and the problem is only libsystemd.so. Think about this: you ship foobard which links to libsystemd.so to both Ubuntu and Fedora users. Ubuntu tells you libsystemd.so is not linked with openssl as that's not allowed. Fedora tells you libsystemd is linked with openssl as that's allowed. They can't both be right, it's the same code with the same licenses. What do you do? This creates doubt and uncertainty, even if things work out in the end. But anyway, looks like we got a good way forward, so it's all moot.

> Here's another idea we had: we currently embed a copy of SHA into sd-boot, because we cannot really link to anything there. It's copied out of glibc, should have OK licensing for all out users hence. Maybe we should move that from sd-boot into src/fundamental/ so that we can compile it for the userspace version too. If we embed a version of that anyway, we might as well use it at multiple places. If we did this we can drop the khash stuff, which Id like very much.

I like this! At that point, if we have this sha256 in use in libsystemd.so instead of khash, would you be ok with dropping the openssl sha256 from it?

bluca commented:
> > I like this! At that point, if we have this sha256 in use in libsystemd.so instead of khash, would you be ok with dropping the openssl sha256 from it?
> 
> So it's a slow implementation, i.e. doesn't use all those fancy CPU features. typically openssl is much preferable hence.
> 
> We could do it like this: use dlopen() to load openssl — witht the RTLD_NOLOAD flag. That way, if the library is already linked to we can just use it, otherwise we'll fall back to the slower implementation we copied from glibc. If people already made the decision that openssl is fine for them to link to from the current process then we'll piggyback on that and also use it. I mean, we use dlopen() stuff all over the place already, this would be just one more place.

Yeah that could work - but are the sd_id128_get_* APIs performance critical and worth the additiona complications?

poettering commented:
note that what we copied is sha256, not more. we need to build hmac around it, but that should be easy. but it should be done with some test vectors that validate that it is correctly implemented.

bluca commented:
All right, let's get this home-grown crypto party going, what could possibly go wrong! I've moved the sha256 over, and implemented hmac following the FIPS spec. Output seems to match openssl's, so it's only ever so slightly possible that it's completely borken.

If we are completely ~~mad~~confident, we could even remove one of the two gcrypt usages from libsystemd, as the journal API in src/libsystemd/sd-journal/journal-authenticate.c uses it to compute sha256 hmacs.

bluca commented:
Just realised sd-id128 was the last usage of khash, so I've nuked it. So long, and thanks for all the hashes!

poettering commented:
lgtm. I'd probably go one step further though and simply define in fundamental/macro.h or so that when building for EFI `typedef UINTN size_t` + `#define memcpy(a, b, c) CopyMem(a, b, c)`.

With that in place we need no ifdeffery in the .c file anymore, and the two identifiers are so well understood that this could not be conflicting with anything.

bluca commented:
> lgtm. I'd probably go one step further though and simply define in fundamental/macro.h or so that when building for EFI `typedef UINTN size_t` + `#define memcpy(a, b, c) CopyMem(a, b, c)`.
> 
> With that in place we need no ifdeffery in the .c file anymore, and the two identifiers are so well understood that this could not be conflicting with anything.

Ok, implemented as requested. Upgrading green label. Thanks!

poettering commented:
lgtm

bluca commented:
focal tests are stuck due to disk space issues on the nodes
